File: bench/stress/s3_vcf.go
// S3 — VCF analysis, Go (std only). go run s3_vcf.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	f, _ := os.Open("data/variants.vcf")
	defer f.Close()
	sc := bufio.NewScanner(f)
	sc.Buffer(make([]byte, 1<<20), 1<<20)

	var total, keep uint64
	counts := make(map[string]uint64)
	for sc.Scan() {
		line := sc.Text()
		if strings.HasPrefix(line, "#") {
			continue
		}
		total++
		c := strings.Split(line, "\t")
		qual, _ := strconv.ParseFloat(c[5], 64)
		if qual > 50.0 {
			keep++
			for _, kv := range strings.Split(c[7], ";") {
				if strings.HasPrefix(kv, "GENE=") {
					counts[kv[5:]]++
				}
			}
		}
	}

	fmt.Printf("total=%d\n", total)
	fmt.Printf("pass=%d\n", keep)
	for _, g := range []string{"BRCA1", "BRCA2", "EGFR", "KRAS", "TP53"} {
		fmt.Printf("%s=%d\n", g, counts[g])
	}
}

File: bench/stress/s3_vcf.rs
// S3 — VCF analysis, Rust (std only). rustc -O s3_vcf.rs -o s3_vcf
use std::collections::HashMap;
use std::fs;

fn main() {
    let raw = fs::read_to_string("data/variants.vcf").unwrap();
    let mut total = 0u64;
    let mut keep = 0u64;
    let mut counts: HashMap<String, u64> = HashMap::new();

    for line in raw.lines() {
        if line.starts_with('#') {
            continue;
        }
        total += 1;
        let c: Vec<&str> = line.split('\t').collect();
        let qual: f64 = c[5].parse().unwrap();
        if qual > 50.0 {
            keep += 1;
            for kv in c[7].split(';') {
                if let Some(rest) = kv.strip_prefix("GENE=") {
                    *counts.entry(rest.to_string()).or_insert(0) += 1;
                }
            }
        }
    }

    println!("total={}", total);
    println!("pass={}", keep);
    for g in ["BRCA1", "BRCA2", "EGFR", "KRAS", "TP53"] {
        println!("{}={}", g, counts.get(g).copied().unwrap_or(0));
    }
}
